apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryDataReader.cs

```diff
@@ -14,197 +14,207 @@
 
 using System;
 using System.Collections;
-using System.Data;
+using System.Collections.Generic;
 using System.Data.Common;
+using System.Threading;
+using System.Threading.Tasks;
+using Google.Api.Gax;
 
 namespace Google.Cloud.BigQuery.V2
 {
     /// <summary>
+    /// Reads a forward-only stream of rows from a data source.
     /// </summary>
     public sealed class BigQueryDataReader : DbDataReader
     {
-#if NET45
+        private readonly BigQueryConnection _connectionToClose;
+        private IAsyncEnumerator<BigQueryRow> _results;
+        private bool _isEnumerating;
 
+        internal BigQueryDataReader(BigQueryResults bigQueryResults, BigQueryConnection connectionToClose = null)
+        {
+            _connectionToClose = connectionToClose;
+            _isEnumerating = true;
+
+            GaxPreconditions.CheckNotNull(bigQueryResults, nameof(bigQueryResults));
+            BigQueryResults = bigQueryResults;
+        }
+
+
+#if NET45 
         /// <inheritdoc />
         public override void Close()
         {
-            throw new NotImplementedException();
+            _results?.Dispose();
+            _connectionToClose?.Close();
         }
 
         /// <inheritdoc />
-        public override DataTable GetSchemaTable()
+        public override System.Data.DataTable GetSchemaTable()
         {
-            throw new NotImplementedException();
+            // We cannot throw here as some scenarios (datatable) will call this method
+            return null;
         }
 #endif
 
         /// <inheritdoc />
-        public override bool NextResult()
-        {
-            throw new NotImplementedException();
-        }
+        public override bool NextResult() => false;
 
         /// <inheritdoc />
         public override bool Read()
-        {
-            throw new NotImplementedException();
-        }
+            => ReadAsync(CancellationToken.None).ResultWithUnwrappedExceptions();
```

Comment:
I see the point of using the async API here - basically we can't mix and match between Read and ReadAsync easily in terms of the use of BigQueryResults.